samcli/lib/providers/sam_api_provider.py

```diff
@@ -261,8 +279,12 @@ def merge_routes(collector):
         all_routes = {}
 
         # By adding implicit APIs to the end of the list, they will be iterated last. If a configuration was already
-        # written by explicit API, it will be overriden by implicit API, just by virtue of order of iteration.
-        all_configs = explicit_routes + implicit_routes
+        # written by explicit API, it will be overridden by implicit API, just by virtue of order of iteration.
+        # Within the explicit/implicit APIs, one defined in top level stack has the higher precedence. Here we
+        # use length of stack_path to order APIs (desc) due to top level stacks always have shorter stack_path.
+        all_configs = sorted(explicit_routes, key=lambda route: -len(route.stack_path)) + sorted(
+            implicit_routes, key=lambda route: -len(route.stack_path)
+        )
```

Comment:
TODO: make it consistent, change the `len(route.stack_path)` to depth (finding how many `/` in there)